shared/src/main/scala/tastyquery/ast/TypeTrees.scala

```diff
@@ -35,45 +37,95 @@ object TypeTrees {
   }
 
   /** ref.type */
-  case class SingletonTypeTree(ref: Tree) extends TypeTree
+  case class SingletonTypeTree(ref: Tree) extends TypeTree {
+    override protected def calculateType: Type =
+      ref.tpe
+  }
 
-  case class RefinedTypeTree(underlying: TypeTree, refinements: List[Tree]) extends TypeTree
+  case class RefinedTypeTree(underlying: TypeTree, refinements: List[Tree]) extends TypeTree {
+    override protected def calculateType: Type =
+      throw new TypeTreeToTypeError(this) // TODO
+  }
 
   /** => T */
-  case class ByNameTypeTree(result: TypeTree) extends TypeTree
+  case class ByNameTypeTree(result: TypeTree) extends TypeTree {
+    override protected def calculateType: Type =
+      ExprType(result.toType)
+  }
 
   /** tpt[args]
     * TypeBounds[Tree] for wildcard application: tpt[_], tpt[?]
     */
-  case class AppliedTypeTree(tycon: TypeTree, args: List[TypeTree | TypeBoundsTree | TypeBounds]) extends TypeTree
+  case class AppliedTypeTree(tycon: TypeTree, args: List[TypeTree | TypeBoundsTree | TypeBounds]) extends TypeTree {
+    override protected def calculateType: Type =
+      AppliedType(
+        tycon.toType,
+        args.map {
+          case arg: TypeTree       => arg.toType
+          case arg: TypeBoundsTree => arg.toTypeBounds
+          case arg: TypeBounds     => arg
+        }
+      )
+  }
 
   /** qualifier#name */
-  case class SelectTypeTree(qualifier: TypeTree, name: TypeName) extends TypeTree
+  case class SelectTypeTree(qualifier: TypeTree, name: TypeName) extends TypeTree {
+    override protected def calculateType: Type =
+      TypeRef(qualifier.toType, name) // TODO Is this correct?
+  }
 
   /** qualifier.TypeName */
-  case class SelectTypeFromTerm(qualifier: Tree, name: TypeName) extends TypeTree
+  case class SelectTypeFromTerm(qualifier: Tree, name: TypeName) extends TypeTree {
```

Comment:
this tree shouldn't really exists, in the tasty is corresponds to some path in a singleton type, so name should be `TermName`, and `calculateType` should be a `TermRef`